sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -147,31 +148,133 @@ class FrameLessOffsetWindowFunctionFrame(
+    }
+  }
+
   override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
     input = rows
     inputIterator = input.generateIterator()
     // drain the first few rows if offset is larger than zero
     inputIndex = 0
-    while (inputIndex < offset) {
-      if (inputIterator.hasNext) inputIterator.next()
-      inputIndex += 1
+    if (ignoreNulls) {
+      findNextRowWithNonNullInput
+    } else {
+      while (inputIndex < offset) {
+        if (inputIterator.hasNext) inputIterator.next()
+        inputIndex += 1
+      }
+      inputIndex = offset
     }
-    inputIndex = offset
+  }
+
+  private val doWrite = if (ignoreNulls && offset > 0) {
+    // For illustration, here is one example: the input data contains six rows,
+    // and the input values of each row are: null, x, null, null, y, null, z, null.
+    // We use Lead(input, 2) with IGNORE NULLS and the process is as follows:
+    // 1. current row -> null, next selected row -> y, output: y;
+    // 2. current row -> x, next selected row -> z, output: z;
+    // 3. current row -> null, next selected row -> z, output: z;
+    // 4. current row -> null, next selected row -> z, output: z;
+    // 5. current row -> y, next selected row -> empty, output: null;
+    // ... next selected row is empty, all following return null.
+    (current: InternalRow) =>
+      if (current.isNullAt(idx)) {
+        if (nextSelectedRow == EmptyRow) {
+          // Use default values since the offset row whose input value is not null does not exist.
+          fillDefaultValue(current)
+        } else {
+          projection(nextSelectedRow)
+        }
+      } else {
+        skipNonNullCount -= 1
+        findNextRowWithNonNullInput
+        if (skipNonNullCount == offset) {
+          projection(nextSelectedRow)
+        } else {
+          // Use default values since the offset row whose input value is not null does not exist.
+          fillDefaultValue(current)
+          nextSelectedRow = EmptyRow
+        }
+      }
+  } else if (ignoreNulls && offset < 0) {
+    // For illustration, here is one example: the input data contains six rows,
+    // and the input values of each row are: null, x, null, null, y, null, z, null.
+    // We use Lag(input, 1) with IGNORE NULLS and the process is as follows:
+    // 1. current row -> null, next selected row -> empty, output: null;
+    // 2. current row -> x, next selected row -> empty, output: null;
+    // 3. current row -> null, next selected row -> x, output: x;
+    // 4. current row -> null, next selected row -> x, output: x;
+    // 5. current row -> y, next selected row -> x, output: x;
+    // 6. current row -> null, next selected row -> y, output: y;
+    // 7. current row -> z, next selected row -> y, output: y;
+    // 8. current row -> z, next selected row -> z, output: z;
+    val absOffset = Math.abs(offset)
+    (current: InternalRow) =>
+      if (nextSelectedRow == EmptyRow && skipNonNullCount == absOffset) {
```

Comment:
We can implement the algorithm with 3 steps:
1. find the next non-null when `skippedNonNullCount == absOffset`
2. set the result row w.r.t. the `nextSelectedRow`
3. check the current row to update `skippedNonNullCount`

The code can be simplied
```
// comment for step 1
if (skipNonNullCount == absOffset) {
  nextSelectedRow = EmptyRow
  while (nextSelectedRow == EmptyRow && inputIndex < input.length) {
    ...
  }
}
// comment for step 2
if (nextSelectedRow == EmptyRow) ... else ...
// comment for step 3
if (!current.isNullAt(idx)) {
  skipNonNullCount += 1
}
```

Reply:
Thank you